lib/ecto/query/builder.ex

```diff
@@ -791,6 +791,84 @@ defmodule Ecto.Query.Builder do
   defp do_literal(value, expected, _current),
     do: {:%, [], [Ecto.Query.Tagged, {:%{}, [], [value: value, type: expected]}]}
 
+  @doc """
+  Escape a values list and extract its parameters
+  """
+  @spec escape_values_list(Macro.t(), Macro.t()) :: {Macro.t(), Macro.t()}
+  def escape_values_list(values_list, types) do
+    fields =
+      quote do
+        Ecto.Query.Builder.values_list_fields(unquote(values_list))
+      end
+
+    types =
+      quote do
+        Ecto.Query.Builder.values_list_types!(unquote(fields), unquote(types))
+      end
+
+    params =
+      quote do
+        Ecto.Query.Builder.values_list_params!(unquote(values_list), unquote(types))
+      end
+
+    num_rows = quote do: length(unquote(values_list))
```

Comment:
It is generally bad practice to unquote the same variable several times, because it means its expression will be executed several times. Perhaps we should encode only once, as parameter, and have all relevant information in the parameter (even if we have to create a struct to represent it).

Reply:
Although I see now that we need the num_rows must be part of the query cache. Perhaps the best option is to ensure on the from/join layer that the values become an `Ecto.Query.Values` struct and that the AST here only points to a variable. Something like this:

    from v in values([...])

becomes:

    internal_ecto_variable = Ecto.Query.Values.new([...])
    from v in values(internal_ecto_variable)

Then in here, you have something like:

    types = quote do: unquote(var).types
    num_rows = quote do: unquote(var).num_rows
    params = quote do: unquote(var).params ++ unquote(params)
    {{:{}, [], [:values, [], [types, num_rows]]}, params}

Although I think you may be able to write it as:

    params = [var | params]

And then in the planner expand the single "var" parameter into many parameters (I believe we do something similar for "in" already).

Reply:
> And then in the planner expand the single "var" parameter into many parameters (I believe we do something similar for "in" already).

I could try playing around with this. The reason I didn't do it initially is because cast/dump/load currently expect the `:in` param to be homogenous in terms of types. 

To make it work we would need to extend it to work with a list of types. And there is a bit of a tradeoff with the error reporting because it's for the entire param and not the individual components. So you'd get something like `can't cast [1, "1", ...] to type [:string, :int, ...]`. I can still play around with it but just wanted to see if you thought this was a non-starter before I do.

Reply:
I see. I am fine with expanding it into multiple parameters here as well. :)

Reply:
I gave it a try with what you said about the struct. The one part I'm unsure of is this:

```elixir
values_struct = quote do: Ecto.Query.Values.new(unquote(values_list), unquote(types))
{expr, params} = Builder.escape_values_list(values_struct)
```

I'm not too sure how to get it to work just passing in the ast for the `values_struct` variable instead of the whole quoted expression with the call to `Ecto.Query.Values.new`. I always get a compiler error `undefined function values_struct/0`.

I was also trying to think if I should be putting it somewhere else. But I think once I'm inside the from/join macros I'll have the same issue wherever I go. If you have any advice for this part I'd appreciate it. Thank you.

Reply:
@greg-rychlewski it needs to be done inside from/join at the same level you do the query. You have to define the Values.new variable before you build the query. Going back to the example, we want this:

    internal_ecto_variable = Ecto.Query.Values.new([...])
    from v in values(internal_ecto_variable)

In think in practice, you can do something like this:

1. Add a new element to From.escape to return `{query :: Macro.t(), prelude :: Macro.t(), params :: Keyword.t()}`

2. For everyone, `:prelude` will simply be `:ok`. For the `values` case, prelude will be `quote do: internal_ecto_variable = Ecto.Query.Values.new([...])`. In this case, you don't need to change `Builder.escape` at all, you can fully handle `values` inside `Builder.From.escape_source`. Something like this:

```elixir
  defp escape_source(query, env) do
    case Macro.expand_once(query, env) do
      {:values, _, [values, types]} = fragment->
        prelude = quote do: values = Ecto.Query.Values.new(unquote(values), unquote(types))
        types = quote do: values.types
        num_rows = quote do: values.num_rows
        params = quote do: values.params ++ unquote(params)
        {{:{}, [], [:values, [], [types, num_rows]]}, prelude, params}

      ... -> ...same code as today...
    end
```

3. Then, in `build`, you will do something like this:

```elixir
    {query, prelude, binds} = escape(query, env)

    query =
      case query do
        ... -> ...same code as today...
      end

    if prelude == :ok do
      query
    else
      quote do
        unquote(prelude)
        unquote(query)
      end
    end
```

I didn't check if this works at all, but that's what I had in mind!

Reply:
Ah I understand now. Thank you! WDYT?

Reply:
for the join one i'm not sure what's better:

1. extract the quoted join expression into a function so the prelude can be conditionally add it
2. keep it as is